docs/edgeql/path_resolution.rst

```diff
@@ -0,0 +1,177 @@
+:ref:`in the Scopes section <ref_eql_path_resolution_scopes>`.
+
+The reason ``with`` works here even though the alias ``U`` refers to the exact
+same set is that we only assume you want the path factored in this way when you
+use the same *symbol* to refer to a set. This means operations with
+``User.first_name`` and ``User.last_name`` *do* get the common path factored
+while ``U.first_name`` and ``User.last_name`` *do not* and are resolved with
+cartesian multiplication.
+
+That may leave you still wondering why ``U`` and ``User`` did not get a common
+path factored. ``U`` is just an alias of ``select User`` and ``User`` is the
+same symbol that we use in our name query. That's true, but EdgeDB doesn't
+factor in this case because of the queries' scopes.
+
+.. _ref_eql_path_resolution_scopes:
+
+Scopes
+------
+
+Scopes change the way path resolution works. Two sibling select queries — that
+is, queries at the same level — do not have their paths factored even when they
+use a common symbol.
+
+.. code-block:: edgeql-repl
+
+    edgedb> select ((select User.first_name), (select User.last_name));
+    {
+      ('Mina', 'Murray'),
+      ('Mina', 'Harker'),
+      ('Mina', 'Westenra'),
+      ('Mina', 'Seward'),
+      ('Jonathan', 'Murray'),
+      ('Jonathan', 'Harker'),
+      ('Jonathan', 'Westenra'),
+      ('Jonathan', 'Seward'),
+      ('Lucy', 'Murray'),
+      ('Lucy', 'Harker'),
+      ('Lucy', 'Westenra'),
+      ('Lucy', 'Seward'),
+      ('John', 'Murray'),
+      ('John', 'Harker'),
+      ('John', 'Westenra'),
+      ('John', 'Seward'),
+    }
+
+Common symbols in nested scopes *are* factored when they use the same symbol.
+In this example, the nested queries both use the same ``User`` symbol as the
+top-level query. As a result, the ``User`` in those queries refers to a single
+object because it has been factored.
+
+.. code-block:: edgeql-repl
+
+    edgedb> select User {
+    ....... name:= (select User.first_name) ++ ' ' ++ (select User.last_name)
+    ....... };
+    {
+      default::User {name: 'Mina Murray'},
+      default::User {name: 'Jonathan Harker'},
+      default::User {name: 'Lucy Westenra'},
+      default::User {name: 'John Seward'},
+    }
+
+If you have two common scopes and only *one* of them is in a nested scope, the
+paths are still factored.
+
+.. code-block:: edgeql-repl
+
+    edgedb> select (Person.name, count(Person.friends));
+    {('Fran', 3), ('Bam', 2), ('Emma', 3), ('Geoff', 1), ('Tyra', 1)}
+
+In this example, ``count``, like all aggregate function, creates a nested
+scope, but this doesn't prevent the paths from being factored as you can see
+from the results. If the paths were *not* factored, the friend count would be
+the same for all the result tuples and it would reflect the total number of
+``Person`` objects that are in *all* ``friends`` links rather than the number
+of ``Person`` objects that are in the named ``Person`` object's ``friends``
+link.
+
```

Comment:
It's possibly worthwhile contrasting this with an example where both parts of that tuple are wrapped into some aggregate function:
```
    edgedb> select (array_agg(distinct Person.name), count(Person.friends));
    {(['Fran', 'Bam', 'Emma', 'Geoff'], 3)}
```
Here `Person` is not factored out because the scopes of aggregate function arguments are _sibling_ scopes, much like the two `select` sub-queries from earlier. So the interpretation is "get an array of all unique names" and "get the total number of people who are somebody's friend".

Reply:
Good suggestion. I added this (nearly verbatim).